File: VideoPoker/Classes/MainView.m
//
//  MainView.m
//  VideoPoker
//
//  Created by Mike McQueen on 8/1/09.
//  Copyright __MyCompanyName__ 2009. All rights reserved.
//

// rename to .mm

//#include "Platform.h"

#import "MainView.h"
#import "UIStrokeLabel.h"

@implementation MainView

@synthesize payScheduleView     = _payScheduleView;
@synthesize returnView          = _returnView;
@synthesize cardTableView       = _cardTableView;

@synthesize dealButton          = _dealButton;
@synthesize holdBestButton      = _holdBestButton;
@synthesize betOneButton        = _betOneButton;
@synthesize betMaxButton        = _betMaxButton;
//@synthesize denominationButton      = _denominationButton;
@synthesize handValueLabel      = _handValueLabel;
@synthesize gameOverLabel       = _gameOverLabel;
@synthesize payScheduleLabel    = _payScheduleLabel;
@synthesize betLabel            = _betLabel;
@synthesize winLabel            = _winLabel;
@synthesize creditsLabel        = _creditsLabel;

- (void)
dealloc
{
    [super dealloc];
}

- (void)
layoutSubviews
{
    /* payScheduleLabel is placed correctly automatically */
    
    /* payScheduleView is placed immediately below payScheduleLabel */
    _payScheduleView.frame = CGRectOffset(_payScheduleView.bounds, 0.0, _payScheduleLabel.frame.origin.y + _payScheduleLabel.frame.size.height + 1);
	
    /* returnView is placed immediately below payScheduleView */
//    CGRect framePSV = _payScheduleView.frame;
//    CGRect boundsRV = _returnView.bounds;
    _returnView.frame = CGRectOffset(_returnView.bounds, 0.0, _payScheduleView.frame.origin.y + _payScheduleView.frame.size.height + 1);
    
    /* cardTableView is placed immediately below returnView */
//    CGRect frameRV = _returnView.frame;
    _cardTableView.frame = CGRectOffset(_cardTableView.bounds, 0.0, _returnView.frame.origin.y + _returnView.frame.size.height + 1);
    
    /* TODO: button layout here! */
}

- (void)
setDealUI
{
    [self.dealButton setTitle:@"DEAL" forState: UIControlStateNormal];
    self.gameOverLabel.hidden = false;
    self.holdBestButton.hidden = true;
    self.betOneButton.hidden = false;
    self.betMaxButton.hidden = false;
}

- (void)
setDrawUI
{
    [self.dealButton setTitle:@"DRAW" forState: UIControlStateNormal];
//    self.handValueLabel.text = nil;
    self.gameOverLabel.hidden = true;
    self.holdBestButton.hidden = false;
    self.betOneButton.hidden = true;
    self.betMaxButton.hidden = true;
    self.winLabel.hidden = true;
}

- (void)
showWin: (uint32_t) payout
{
    self.winLabel.text = [NSString stringWithFormat: @"WIN %d", payout];
    self.winLabel.hidden = false;
}

- (void)
showCash: (double) cash
{
    self.creditsLabel.text = [NSString stringWithFormat: @"CASH $%.2f", cash];
    self.creditsLabel.hidden = false;
}

/*
- (id)
initWithFrame: (CGRect) frame
{
    if (self = [super initWithFrame:frame]) 
	{
        // Initialization code
    }
    return self;
}

- (void)
drawRect: (CGRect) rect
{
    // Drawing code
}
*/

@end

File: VideoPoker/Classes/MainView.swift
//
//  MainView.swift
//  VideoPoker
//
//  Created by Mike McQueen on 8/1/09.
//

import UIKit

/// The Train tab: pay table, returns, cards, and the controls below them
final class MainView: UIView {
    var payScheduleView: UIView?
    var returnView: UIView?
    var cardTableView: UIView?

    @IBOutlet var dealButton: UIButton!
    @IBOutlet var holdBestButton: UIButton!
    @IBOutlet var betOneButton: UIButton!
    @IBOutlet var gameOverLabel: UILabel!
    @IBOutlet var handValueLabel: UILabel!
    @IBOutlet var payScheduleLabel: UILabel!
    @IBOutlet var betLabel: UILabel!
    @IBOutlet var winLabel: StrokeLabel!
    @IBOutlet var creditsLabel: StrokeLabel!

    /// Pay table height the controls below the card table were laid out for in the nib
    private static let nibPayScheduleHeight: CGFloat = 140.0
    /// Nib y positions of the controls below the card table
    private var controlBaseY: [CGFloat]?

    private var controlsBelowCards: [UIView] {
        [dealButton, holdBestButton, betOneButton, gameOverLabel, handValueLabel, betLabel, winLabel, creditsLabel]
    }

    override func layoutSubviews() {
        // payScheduleLabel is placed correctly automatically; stack the rest below it
        guard let payScheduleView = payScheduleView, let returnView = returnView, let cardTableView = cardTableView else {
            return
        }
        payScheduleView.frame = payScheduleView.bounds.offsetBy(dx: 0.0, dy: payScheduleLabel.frame.maxY + 1)
        returnView.frame = returnView.bounds.offsetBy(dx: 0.0, dy: payScheduleView.frame.maxY + 1)
        cardTableView.frame = cardTableView.bounds.offsetBy(dx: 0.0, dy: returnView.frame.maxY + 1)

        // Shift the controls below the card table by however much the pay table grew
        let controls = controlsBelowCards
        let baseY = controlBaseY ?? controls.map { $0.frame.origin.y }
        controlBaseY = baseY
        let delta = payScheduleView.bounds.height - MainView.nibPayScheduleHeight
        for (control, y) in zip(controls, baseY) {
            control.frame.origin.y = y + delta
        }
    }

    func setDealUI() {
        dealButton.setTitle("DEAL", for: .normal)
        gameOverLabel.isHidden = false
        holdBestButton.isHidden = true
        betOneButton.isHidden = false
    }

    func setDrawUI() {
        dealButton.setTitle("DRAW", for: .normal)
        gameOverLabel.isHidden = true
        holdBestButton.isHidden = false
        betOneButton.isHidden = true
        winLabel.isHidden = true
    }

    func showWin(_ payout: Int) {
        winLabel.text = "WIN \(payout)"
        winLabel.isHidden = false
    }

    func showCash(_ cash: Double) {
        creditsLabel.text = String(format: "CASH $%.2f", cash)
        creditsLabel.isHidden = false
    }
}
